src/main/java/com/hubspot/jinjava/lib/tag/eager/EagerSetTag.java

```diff
@@ -0,0 +1,133 @@
+package com.hubspot.jinjava.lib.tag.eager;
+
+import com.hubspot.jinjava.interpret.Context;
+import com.hubspot.jinjava.interpret.DeferredValueException;
+import com.hubspot.jinjava.interpret.JinjavaInterpreter;
+import com.hubspot.jinjava.interpret.TemplateSyntaxException;
+import com.hubspot.jinjava.lib.tag.DoTag;
+import com.hubspot.jinjava.lib.tag.SetTag;
+import com.hubspot.jinjava.tree.parse.TagToken;
+import com.hubspot.jinjava.util.ChunkResolver;
+import java.util.Arrays;
+import java.util.List;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+
+public class EagerSetTag extends EagerStateChangingTag<SetTag> {
+
+  public EagerSetTag() {
+    super(new SetTag());
+  }
+
+  public EagerSetTag(SetTag setTag) {
+    super(setTag);
+  }
+
+  @Override
+  public String getEagerTagImage(TagToken tagToken, JinjavaInterpreter interpreter) {
```

Comment:
A more general question here:

You probably understand this better than me at this point, but why don't we implement this by manipulating the tree directly?

The way most compilers work is by parsing the code up into a tree (which we have), doing n passes through the tree to manipulate nodes and perform evaluation when possible. Then they write out the code in one pass. Could we do that here instead of returning a string as the output?

Reply:
The jinjava engine is an interpreter rather than a compiler so it parses one line at a time into its output. Not lines exactly, as it first parses the template into a tree, and then it interprets each node in that tree as it sees it (because unlike python, jinja doesn't have a strong line-based syntax). 

The way I have this implemented is so that nodes are still evaluated on the fly, but their evaluation is handled differently when a value is deferred. Rather than throwing an error there, a specific output can be generated such that running that output with resolved values will produce the same output as running the original with resolved values, but faster. This approach allows for the sequential node evaluation to be done as normal.

In a different rendering engine that's more c-based than python-based, it would probably be done by manipulating the tree